Turn On Red (Previously Right On Red)

FireController1847 opened the pull request:
Reference to https://github.com/VictorPhilipp/Cities-Skylines-Traffic-Manager-President-Edition/pull/114

Finally learned how to do this. Separated it from all the other stuff in that PR, and then realised the way it was being done was stupid so did it a better way lol

Sorry 'bout my formatter (I just used default VS2017). In the diffs it looks super off center but in Visual Studio it looks fine (different between tabs & spaces)

But here it is, ready to go!

I also snuck in another debug mode for resources, that way we can understand if there's a missing resource or something (when a resource doesn't load, it crashes the entire menu so it doesn't show up. Spent hours debugging just to find out that a resource wasn't loading).

Edit: Yay, more bugs. Turn-on-red is not working at all for certain types of vehicles. Unsure why. Will investigate.

krzychu124 commented:
Is it bug less? 😄 
I will try to find some time tomorrow to look at this. 
I think I've seen this as separate worshop mod

FireController1847 commented:
> Is it bug less? 😄
> I will try to find some time tomorrow to look at this.
> I think I've seen this as separate worshop mod

I have been playing with it for a while now and it seems to work well. However there are a few things to be done (see the TODO comments). I've seen the separate workshop mod as well, however I feel like this is 1) much more fitting for this mod and 2) I was unable to get the other mod to work with TM;PE.

krzychu124 commented:
You mean this? 
>// TODO: If is a wreckless driver, yield / run
                                // TODO: Disallow cutting into flowing traffic, disrupting the flow.
                                // TODO: If turning onto a one-way road, allow turning for the opposite direction.
Would be nice to resolve some of them before merging :)

FireController1847 commented:
I'll try and find some time this week to see if I can't resolve some of those. However with some of those I may be unable to, since I'm still learning how this mod works in its entirety. I'm pretty sure one-way roads won't be that hard, same with wreckless drivers. The cutting into the flow of traffic might be a little harder though. I'll take my best shot when I get the chance

krzychu124 commented:
Don't worry I'm still learning too. Codebase of this game is huge and sometimes complicated :)
BTW I will try to add wiki page in which I'll describe how to attach debugger to running game and place breakpoints at any line 😄

FireController1847 commented:
> Don't worry I'm still learning too. Codebase of this game is huge and sometimes complicated :)

It is _extremely_ complicated. Like, why are there over 15 classes just to control chirper? And WHY are events tied so deeply into the messages for chirper? Even with that, there's no generic list of messages for each event? And that's just one example.

> BTW I will try to add wiki page in which I'll describe how to attach debugger to running game and place breakpoints at any line 😄

That sounds awesome, I'd love to see an example of how to debug and place breakpoints. Heh, I spent more time than I'd like to admit figuring out that it logs to a file called "TMPE.log" in the game directory, as well as how to enable debugging.

FireController1847 commented:
I've fixed two of the three checks. I might be able to work on the other ones tomorrow, and I've got an idea on how to do it.

FireController1847 commented:
No worries! I'll also keep digging and see if there's anything I've missed while looking through the code.

krzychu124 commented on TLM/TLM/Resources/lang.txt:
```diff
@@ -151,6 +151,7 @@ General General
 Gameplay Gameplay
 Overlays Overlays
 Realistic_speeds Realistic speeds
+Turn_on_red Turn on red
```
Would be good to copy that line to all lang files

krzychu124 commented on TLM/TLM/Manager/Impl/VehicleBehaviorManager.cs:
```diff
@@ -334,6 +335,47 @@ public class VehicleBehaviorManager : AbstractCustomManager, IVehicleBehaviorMan
+
+                                        // If the direction you are going is the same as the preferred lane or you're going from a one-way road to another one-way road, turn
+                                        SegmentGeometry turnSegGeo = SegmentGeometry.Get(uTurnSegment);
+                                        if (uTargetSegment == uTurnSegment || (currentSegGeo.IsOneWay() == true && turnSegGeo?.IsOneWay() == true)) {
```
I think you can omit that first comparator with 'true' 😉 
 ```cs
if (uTargetSegment == uTurnSegment || (currentSegGeo.IsOneWay() && turnSegGeo?.IsOneWay() == true)) {
```

FireController1847 replied:
Wow, I didn't even notice that. Good for readability 👍

krzychu124 commented:
Another issue 😉 As you can see it would be great to hide this sign if there is no right turn 😄 
![zrzut ekranu 50](https://user-images.githubusercontent.com/19638970/52365542-0320bd80-2a48-11e9-8546-d6c93185c4ff.png)
Would be nice to see this feature in upcoming patch 1.10.15 😉

krzychu124 commented:
https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/33#issuecomment-461152346 
I've added my thoughts about where this feature should have been located 
(toggle button to enable/disable).

FireController1847 commented:
Hopefully the latest commit solves almost all of your review requests. I've also created some cool looking signs for this. [Here's the PSD files I used to make them](https://1drv.ms/f/s!AlmdHdoNvQvRgvcHQMJOG0PbCdm-uA), as it provides an easy template for making future signs as well. I found the font that is used in the SOS sign, and it turns out that font has a BUNCH of cool road signs on it. It's called, "Roadgeek 2005." You should search it up, it's pretty neat if you ever need to create some more road signs.

On another note, I've noticed that yield signs do exactly what the TODO is having an issue with; detecting right-of-way. I'll take a look at that soon and see how it detects if there's another car in the lane (as when yield signs are enabled, cars do not turn unless it is clear).

FireController1847 commented on TLM/TLM/State/Options.cs:
```diff
@@ -1187,6 +1198,12 @@ public class Options : MonoBehaviour {
 
+        public static void setTurnOnRed(bool newValue) {
+            turnOnRed = newValue;
+            if (turnOnRed != null)
```
Whoops! Yep, my bad.

FireController1847 commented:
Alright. Other than one final bug against some vehicles not turning on red (the same ones that stop at green lights before continuing, like wtf is up with that? trucks I've noticed do this), this is ready. Wohoo!

krzychu124 commented:
I've noticed that behavior of that vehicle stop before continuing. 
<table border="0">
<tr>
<td>In Poland you have to do this if main light is red and there is conditional green arrow btw</td>
<td><img src="https://image.shutterstock.com/display_pic_with_logo/3990793/508868143/stock-photo-red-traffic-light-with-a-green-arrow-pedestrian-crossing-and-give-way-yield-signs-508868143.jpg"></td>
</tr>
</table>

Haha, I've just noticed you didn't just copy & paste description but you've translated to every other language 😄

pcfantasy commented:
Hi, All:
The separate workshop mod is written by me. And I will update that mod to support

1. Turn on red
2. U-Turn on red
3. Left-waiting when straight light is green but left light is red

I think (1)and (2) can be merged into TMPE for better use.

github code is in
https://github.com/pcfantasy/AdvancedJunctionRule

Can you guys also add U-turn and left-waiting function to this mod?

FireController1847 commented:
@pcfantasy As far as I know, TMPE already has u-turn support, as well as left-waiting. If you set up manual traffic lights it should work like that (hit Change Mode 'til you see the left arrow, then when the straight is green and the left arrow is red it ends). Also, with #3, if you mean while the light is green to wait for clear traffic, I've _never_ seen that before. Is that local to a different country? Over here, we call it "dedicated left" because if the light is green and is a left arrow, the opposite light is always red. For turn on red, it is most likely going to be added as of this PR.

@krzychu124 That's very interesting. Although when I was talking about going through green, I meant they were stopping at green lights going straight, that confused me. Seems pretty rare though.

pcfantasy commented:
@FireController1847 

I know it is hard for you to understand left waiting function, it is popular for east asia country, like China and Japan.

The function is, when straight light is green and left light is red, left turn cars will enter a little bit in junction and wait for next phase(straight light is red and left light is green). 


And for U-turn on red, in most case  U-turn is allowed only when left light is green, but sometimes in some junctions, when straigt light and left light is red, U-turn actually will not impact on the traffic, so we can add this function.

FireController1847 commented:
@pcfantasy I see. I feel like these are features for another PR, possibly. Good ideas all round, though!

krzychu124 commented:
@FireController1847 with regards to stop at green lights I've noticed that too. 
This is a current limitation of checkSpace function. 
If you look closely you will see that cars stop at green lights only when next segment is almost full and in their current simulation step were no space to fit in 
(it's fight between flags leave and blocked )

aubergine10 commented:
I've just referenced that video of "enter when waiting" in an issue I created to track this PR: https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/46

It's exactly what traffic in UK does, only on other side (we drive on left of road).

FireController1847 commented:
OH! Now I understand what you guys are talking about! Look up "yellow flashing left turn light", it has similar functionality.

krzychu124 commented:
@FireController1847 any progress?

FireController1847 commented:
It think it should be ready to merge as it is now, unless it will break saving junctions (ref. #32)

krzychu124 commented:
Ok, I'm currently checking that situation ;)

krzychu124 commented:
More issues 😢 
* one-way bend

![zrzut ekranu 64](https://user-images.githubusercontent.com/19638970/52499065-e23ca180-2bda-11e9-8dae-0fb44aa22ca8.png)

* sometimes right turn availability not calculated correctly

![zrzut ekranu 66](https://user-images.githubusercontent.com/19638970/52499066-e2d53800-2bda-11e9-9694-263abc0b79ac.png)

* level crossings - I think you can disable for any

![zrzut ekranu 68](https://user-images.githubusercontent.com/19638970/52499067-e2d53800-2bda-11e9-9b4c-4faf130498ae.png)

I hope that's all, but I will continue testing

FireController1847 commented:
Any particular reason we'd disable it by default for every junction? If you want it disabled entirely, wouldn't you just not use the option? The way I see it, the only time you'd enable the option is if you wanted it. In that case, it'd be a huge pain to enable it for every junction _every time you made a junction_, versus just disabling when you didn't want it. I understand Europeans and countries that don't allow this, but in that case they'd just disable the option no?

krzychu124 commented:
Let's say I would like to _enable_ this feature only for selected junctions to simulate conditional right turn. Unfortunately with your current implementation it's almost impossible. 

The first thought that comes to my mind is to add another toggle to leave decision to user how to interpret current settings(enable or disable all). 
I think it's just one more bool variable to flip functionality that will save a lot of reserved space for mod configuration inside savegame in my case. 
What do you think?

FireController1847 commented:
Ah, I see. Sure, one more toggle switch isn't much, I can add that. I'm going to dinner now, so I'll be back in about an hour and a half then I'll be able to get it setup on my laptop and make those changes.

FireController1847 commented:
Alright, so I've added that toggle switch and it appears to be working. I've also made a change to the detection logic. I accidentally left in a mistake I made a while back. Basically, it was enabling it if there was a "Oneway Out" in ALL cases. I also added better detection for one-way roads to prevent possible future bugs, and I added a conditional for railroad crossings.

I appreciate you testing this for me, as I've been playing with it casually on but I'm not really good at coming up with random things like a buslane road onto a highway, or a road onto train tracks. I'm happy to help out and fix the bugs. After these next two commits it should be ready to merge, however I said that before and we found more bugs 😅 

On a random note, if you're interested, [this resource](https://le.utah.gov/xcode/Title41/Chapter6a/41-6a-S305.html#41-6a-305(4)(a)) has some interesting things about red lights, particularly in my state of Utah. 

(A) | The operator of a vehicle facing a steady circular red signal may cautiously enter the intersection to turn right, or may turn left from a one-way street into a one-way street, after stopping as required by Subsection (4)(a).

(B) | If permitted by a traffic control device on the state highway system, the operator of a vehicle facing a steady red arrow signal may cautiously enter the intersection to turn left from a one-way street into a one-way street after stopping as required by Subsection (4)(a).

aubergine10 commented:
For the whole "cautiously entering junction" thing, maybe we need 3 states for the 'blocked junction' icon:

* Don't enter blocked junctions (see [yellow box junction](https://tfl.gov.uk/modes/driving/red-routes/rules-of-red-routes/yellow-box-junctions))
* Cautiously enter for left/right turn, as applicable to driving direction (see [this video](https://www.youtube.com/watch?v=ICjrLoy6tFA))
* Allow enter blocked junction (see: India, Egypt...)

krzychu124 commented:
Level crossings and that bend one-way from screenshots I've posted were fixed but I can see that there is still problem with detecting direction of outgoing one-way segments 😭
I will try to check what is going on there.

Btw. I've fixed that null pointer exception just after game intro loading screen

FireController1847 commented:
@krzychu124 I tested with it just a little bit, and if you put a one-way road going in the opposite direction directly behind the intersection of the one-way road turn, it fixes it. Invalid segments, possibly? Clearly it has an issue with my `HasPreferredSegment` logic, but I'm unsure of where.

krzychu124 commented on TLM/TLM/Geometry/Impl/SegmentGeometry.cs:
```diff
@@ -898,13 +898,29 @@ public class SegmentGeometry : IEquatable<SegmentGeometry> {
+            bool hasLeftSegStart = HasLeftSegment(true);
+            bool hasLeftSegEnd = HasLeftSegment(false);
+            return (
+                ((Constants.ServiceFactory.SimulationService.LeftHandDrive ? hasLeftSegStart || hasLeftSegEnd : hasRightSegStart || hasRightSegEnd)) ||
```
Is it possible to simplify this a little bit because I am not sure I understand what is going here 😄

FireController1847 replied:
Quite frankly me either. I don't know the different between the startNode being true and false. Basically, what's happening is it's searching for if a segment has a segment to its right or to its left. If it does, it stores it in a bool, then I use those books to compare. The reason I check both the start and the end is that I've noticed it works much better than when I only do the end and/or only do the start, however I could probably do it better if I knew the difference lol

krzychu124 replied:
I will try to refactor this a little bit because I think I know how it should look like. Give me some time 😉

krzychu124 replied:
Ok, I think I've fixed all issues after refactoring.

krzychu124 replied:
instead of HasPreferredSegment() add this (it has more descriptive name what it's doing):
```cs
/// <summary>
/// Determines if, according to the stored geometry data, the segment is eligible for turning right-on-red.
/// 
/// A segment geometry verification is not performed. 
/// </summary>
/// <param name="segmentEndGeometry">segment end geometry of incoming segment to node</param>
/// <returns></returns>
public bool HasValidTurnOnRedOutgoingSegment(SegmentEndGeometry segmentEndGeometry) {
    bool isLeftHandDrive = Constants.ServiceFactory.SimulationService.LeftHandDrive;
    return isLeftHandDrive ? segmentEndGeometry.NumOutgoingLeftSegments > 0 : segmentEndGeometry.NumOutgoingRightSegments > 0;
}
```
and in JunctionRestrictionsManager.cs delete lines 269 and 270 and replace with this:
```cs
segGeo?.HasValidTurnOnRedOutgoingSegment(endGeo) == true;
```

As you can see we don't have to check if its one-way, I think

krzychu124 replied:
And of course lines 352 and 354 in VehicleBehaviorManager.cs 😉

FireController1847 replied:
I can't tell exactly where, but from what I know it also needs to detect if it's a left-to-left from a one-way to a one-way, as vehicles may turn left on red if it's from a one-way road to another one-way road.

FireController1847 replied:
Thank you for looking at that! I'm glad you found a better way to do it, it seems to be much more stable now. Apologies I wasn't able to figure this out before >.<

krzychu124 replied:
Ok, I can see this now in my game..
![notitle](https://user-images.githubusercontent.com/19638970/52525932-8bee6200-2cb1-11e9-9d9c-6623737388cd.jpg)

FireController1847 replied:
I'm noticing they're not actually performing the right on red anymore either. Why is it that every time we fix this, it breaks again? 😫 

I'll take another look for the 9 billionth time and maybe see if I can't make a major change to the way it actually performs it.

FireController1847 commented:
@aubergine10 Also, look up the "left turn yield at Green" sign. It had the same functionality as flashing yellow left.

krzychu124 commented on TLM/TLM/Manager/Impl/VehicleBehaviorManager.cs:
```diff
@@ -347,11 +347,12 @@ public class VehicleBehaviorManager : AbstractCustomManager, IVehicleBehaviorMan
 
                                     // If you can turn preferred and you're not going straight, continue
                                     SegmentGeometry currentSegGeo = SegmentGeometry.Get(uCurrentSegment);
+                                    SegmentEndGeometry currentSegEndGeo = currentSegGeo.EndNodeGeometry;
```
Are you sure it's EndNodeGeometry?

krzychu124 replied:
```suggestion
                                  SegmentEndGeometry currentSegEndGeo = currentSegGeo.GetEnd(targetNodeId);
```

krzychu124 replied:
For me, with this it still working as expected

FireController1847 replied:
KMS!

Okay, I have to ask this. What is the difference between a Segment and a SegmentEnd? Also, what does this "startNode" option do/mean (if I do `currentSegGeo.GetEnd(startNode);`, I need the startNode property).

krzychu124 replied:
Segment has obviously two ends (connected nodes) so GetEnd(node) will get you proper end (imagine that you are in that vehicle approaching node - now you can easily decide which is left and right outgoing segment)

FireController1847 commented:
Okay, okay. I set up a _bunch_ more test situations, and I think this latest commit gets most of them, hopefully?

krzychu124 commented:
If we tackle down that issue from my last screenshot I will push 1.10.15 to Steam Workshop 😉

krzychu124 commented on TLM/TLM/Manager/Impl/VehicleBehaviorManager.cs:
```diff
@@ -347,7 +347,7 @@ public class VehicleBehaviorManager : AbstractCustomManager, IVehicleBehaviorMan
                                     // If you can turn preferred and you're not going straight, continue
                                     SegmentGeometry currentSegGeo = SegmentGeometry.Get(uCurrentSegment);
-                                    SegmentEndGeometry currentSegEndGeo = currentSegGeo.EndNodeGeometry;
+                                    SegmentEndGeometry currentSegEndGeo = currentSegGeo.GetEnd(true);
```
Am I seeing correctly? currentSegGeo.GetEnd(true)? 
What is that 'true' and more interesting how it's even casted to type ushort 😆

FireController1847 replied:
GetEnd asks for a bool of startNode. I am confused?

krzychu124 replied:
`(+1 overload)` 😉

FireController1847 commented:
> If we tackle down that issue from my last screenshot I will push 1.10.15 to Steam Workshop 😉

That should be fixed in the latest commit, no?
![20190209135606_1](https://user-images.githubusercontent.com/16494272/52526220-bec53580-2c72-11e9-97e4-1ce61fd5e7b4.jpg)

FireController1847 commented:
Please tell me I did it right this time lol...

Luckily, in practice, every intersection I have seems to be working well.

krzychu124 commented:
Hahaha, I see you have identical code to mine now but I am still not sure if it's working as intended 🤣

krzychu124 commented:
Ok, I'm going to check what is going on with this left-left/right-right one-way

krzychu124 commented:
I think that option 'Turn on red enabled at junctions by default' is not taken into account while drawing sings

FireController1847 commented:
As far as I know, that rule can only apply to brand new junctions. It'd be very difficult to change all the current junctions while still keeping the ones manually changed, no?

krzychu124 commented:
Maybe you should just flip state while drawing?

FireController1847 commented:
I don't know, I feel like that could be very confusing. But sure, that's pretty easy, I think

krzychu124 commented:
Argh... it still not working properly but now in other place. 😭  I think I don't understand that segment ends too... I hope there is no bug up there 😄

FireController1847 commented:
Which place is it not working? I've tried many different roads and it's working for me

krzychu124 commented:
Try this: (right hand driving) left and straight are one-way outgoing

![zrzut ekranu 76](https://user-images.githubusercontent.com/19638970/52526769-6e73c500-2cbe-11e9-9d8c-6dbfc6f2155f.png)

FireController1847 commented:
That is the intended behavior. If turning from a one-way road to another one-way road, you are allowed to turn left at red.

krzychu124 commented:
Isn't it make cutting other vehicles way? Say I am at right lane, as you can see I can't even turn left (lane arrows) It's confusing to me 😕

FireController1847 commented:
I guess from the outer lane, it would technically be cutting into the inner lane. Here's some references.

- [The google search I used](https://www.google.com/search?q=one-way+road+to+another+one-way+road+left+turn+on+red&oq=one-way+road+to+another+one-way+road+left+turn+on+red)
- Left turn from a one-way street into a one-way street. Start the turn from the far left lane. Watch for pedestrians, motorcyclists, and bicyclists between your vehicle and the curb because they can legally use the left turn lane for their left turns. Turn into any lane that is safely open, as shown by the arrows. ([ref 1](https://www.dmv.ca.gov/portal/dmv/detail/pubs/hdbk/turns))
- Drivers also can make a left turn onto a one-way street, unless a posted sign prohibits the movement. The same rules apply: First, come to a complete stop at the red light, and give the right of way to other vehicles in or approaching the intersection. ([ref 2](https://www.bellinghamherald.com/news/traffic/rules-of-the-road/article87261507.html))

aubergine10 commented:
Have you tested this on roads with bus and bike lanes?

Also, in terms of whether to stop or yield, does it need integrating with the stop/yield/priority signs?

FireController1847 commented:
@aubergine10 No, this only applies to traffic lights.
@krzychu124 In your edited example, since it is not turning from a one-way into another one-way road, it should not be an option. Is it an option in that example?

krzychu124 commented:
Ok, I understand this and yes option to turn right is visible 😕

![zrzut ekranu 80](https://user-images.githubusercontent.com/19638970/52526967-5baebf80-2cc1-11e9-93c6-283bc9c78cfa.png)

* Technically if we look closely that outgoing straight road could be registered as right and I have no idea how to hide that sign in this case...

FireController1847 commented:
You know, what? I gotchya, mate! Haha, I have no idea why this worked but it did. (It's the same logic as before, just #1 separated and #2 adding a check if the opposing direction is a OutgoingOneWay road.

krzychu124 commented:
Yup it's working now. Merging, after almost 80 comments 😄

FireController1847 commented:
Hey, you know, it's worth it to take the time a bugtest! Glad we got it to work!

krzychu124 commented:
I know but it's not master branch yet 😉

krzychu124 commented:
I think we've forgot about tram tracks...

FireController1847 commented:
That's.. unfortunate. I'll check out.

aubergine10 commented:
Would it be worth releasing the existing stuff (fixed Est, localisations, icons) to Steam so people have some improvements, and release the Turn on right/left thing as a separate beta mod to allow community to heavily test it (similar to how Loading Screen Mod has two workshop pages)?

FireController1847 commented:
Well, actually, it seems that tram tracks also have issues with u-turn and other restrictions. I don't think it's limited to right turns.

krzychu124 commented:
Nonetheless it's worth fixing, but I think it need new issue
With regards to releasing new version I think it's possible and I will add (Beta) to new toggles in options menu 😄
